File: shacs_bot/agent/tools/cron/cron.py
"""알림과 작업 스케줄링을 위한 Cron 도구"""

from _contextvars import ContextVar, Token
from datetime import datetime
from typing import Any
from zoneinfo import ZoneInfo

from shacs_bot.agent.tools.base import Tool
from shacs_bot.agent.tools.cron.service import CronService
from shacs_bot.agent.tools.cron.types import CronSchedule, CronJob


class CronTool(Tool):
    """스케줄 알림과 반복 작업을 위한 도구입니다."""

    name = "cron"
    description = "알림과 반복 작업을 스케줄링합니다. Actions: add, list, remove."
    parameters = {
        "type": "object",
        "properties": {
            "action": {
                "type": "string",
                "enum": ["add", "list", "remove"],
                "description": "수행할 작업",
            },
            "message": {"type": "string", "description": "알림 메시지 (추가 시)"},
            "every_seconds": {"type": "integer", "description": "반복 작업의 간격(초 단위)"},
            "cron_expr": {
                "type": "string",
                "description": "'0 9 * * *'와 같은 크론 표현식(예약된 작업 시)",
            },
            "at": {
                "type": "string",
                "description": "일회성 알림의 실행 시각 (ISO 형식: YYYY-MM-DDTHH:MM:SS). 지정 시간에 한 번 실행 후 자동 삭제됩니다.",
            },
            "job_id": {"type": "string", "description": "작업 ID(제거 시)"},
        },
        "required": ["action"],
    }

    def __init__(self, cron_service: CronService):
        self._cron: CronService = cron_service
        self._channel: str = ""
        self._chat_id: str = ""
        self._in_cron_context: ContextVar[bool] = ContextVar("cron_in_context", default=False)

    def set_context(self, channel: str, chat_id: str) -> None:
        """전달을 위한 현재 세션 컨텍스트를 설정합니다."""
        self._channel: str = channel
        self._chat_id: str = chat_id

    def set_cron_context(self, active: bool) -> Token[bool]:
        """도구가 cron 작업 콜백 내부에서 실행 중인지 여부를 표시한다."""
        return self._in_cron_context.set(active)

    def reset_cron_context(self, token: Token[bool]) -> None:
        """이전 cron 컨텍스트 복구"""
        self._in_cron_context.reset(token)

    async def execute(
        self,
        action: str,
        message: str = "",
        every_seconds: int | None = None,
        cron_expr: str | None = None,
        tz: str | None = None,
        at: str | None = None,
        job_id: str | None = None,
        **kwargs: Any,
    ) -> str:
        if action == "add":
            if self._in_cron_context.get():
                return "에러: cron 작업 실행 중에는 새로운 작업을 예약할 수 없습니다."

            return self._add_job(
                message=message, every_seconds=every_seconds, cron_expr=cron_expr, tz=tz, at=at
            )
        elif action == "list":
            self._list_jobs()
        elif action == "remove":
            return self._remove_job(job_id)

        return f"알 수 없는 작업: {action}"

    def _add_job(
        self,
        message: str,
        every_seconds: int | None,
        cron_expr: str | None,
        tz: str | None = None,
        at: str | None = None,
    ) -> str:
        if not message:
            return "에러: 추가를 위해 메시지가 필요합니다."

        if not self._channel or not self._chat_id:
            return "에러: 세션 컨텍스트가 설정되지 않았습니다 (channel/chat_id)."

        if tz and not cron_expr:
            return "에러: tz는 cron_expr와 함께 사용할 때만 가능합니다."

        if tz:
            try:
                ZoneInfo(tz)
            except (KeyError, Exception):
                return f"에러: 알 수 없는 시간대 '{tz}'"

        # 스케줄 추가
        delete_after = False
        if every_seconds:
            schedule: CronSchedule = CronSchedule(kind="every", every_ms=every_seconds * 1000)
        elif cron_expr:
            schedule = CronSchedule(kind="cron", expr=cron_expr)
        elif at:
            try:
                dt: datetime = datetime.fromisoformat(at)
            except ValueError:
                return (
                    f"에러: 잘못된 ISO datetime 형식입니다 '{at}'. 예상 형식: YYYY-MM-DDTHH:MM:SS"
                )

            at_ms: int = int(dt.timestamp() * 1000)
            schedule: CronSchedule = CronSchedule(kind="at", at_ms=at_ms)

            delete_after = True
        else:
            return "에러: every_seconds, cron_expr, at 중 하나는 필요합니다."

        job: CronJob = self._cron.add_job(
            name=message[:30],
            schedule=schedule,
            message=message,
            deliver=True,
            channel=self._channel,
            to=self._chat_id,
            delete_after_run=delete_after,
        )

        return f"작업이 추가되었습니다: '{job.name}' (id: {job.id}"

    def _list_jobs(self) -> str:
        jobs: list[CronJob] = self._cron.list_jobs()
        if not jobs:
            return "예약된 작업이 없습니다."

        lines: list[str] = [f"- {job.name} (id: {job.id}, {job.schedule.kind})" for job in jobs]
        return "스케줄된 작업:\n" + "\n".join(lines)

    def _remove_job(self, job_id: str | None) -> str:
        if not job_id:
            return "에러: 제거를 위해 job_id가 필요합니다."

        if self._cron.remove_job(job_id):
            return f"작업이 제거되었습니다: {job_id}"

        return f"작업을 찾을 수 없습니다: {job_id}"

File: crates/shacs-core/src/tools/cron.rs
use crate::tools::{BooleanSchema, IntegerSchema, JsonMap, StringSchema, Tool, ToolParameters};
use crate::tools::{SchemaFragment, ToolResult, ValidationError};
use chrono::{DateTime, LocalResult, NaiveDateTime, TimeZone};
use chrono_tz::Tz;
use serde_json::Value;
use shacs_cron::{
    AddCronJob, CronJob, CronJobState, CronPayloadKind, CronRunStatus, CronSchedule,
    CronScheduleKind, CronService, RemoveJobResult,
};
use std::cell::RefCell;
use std::sync::Arc;

thread_local! {
    static CRON_CONTEXT: RefCell<CronContext> = RefCell::new(CronContext::default());
}

#[derive(Debug, Clone)]
struct CronContext {
    channel: String,
    chat_id: String,
    metadata: Value,
    session_key: String,
    in_cron_context: bool,
}

impl Default for CronContext {
    fn default() -> Self {
        Self {
            channel: String::new(),
            chat_id: String::new(),
            metadata: Value::Object(Default::default()),
            session_key: String::new(),
            in_cron_context: false,
        }
    }
}

#[derive(Clone)]
pub struct CronTool {
    cron: Arc<dyn CronService>,
    default_timezone: String,
}

impl CronTool {
    pub fn new(cron: Arc<dyn CronService>) -> Self {
        Self::with_timezone(cron, "UTC")
    }

    pub fn with_timezone(cron: Arc<dyn CronService>, default_timezone: impl Into<String>) -> Self {
        Self {
            cron,
            default_timezone: default_timezone.into(),
        }
    }

    pub fn set_context(
        &self,
        channel: impl Into<String>,
        chat_id: impl Into<String>,
        metadata: Option<Value>,
        session_key: Option<String>,
    ) {
        let channel = channel.into();
        let chat_id = chat_id.into();
        CRON_CONTEXT.with(|context| {
            let mut context = context.borrow_mut();
            context.channel = channel.clone();
            context.chat_id = chat_id.clone();
            context.metadata = metadata.unwrap_or_else(|| Value::Object(Default::default()));
            context.session_key = session_key.unwrap_or_else(|| format!("{channel}:{chat_id}"));
        });
    }

    pub fn set_cron_context(&self, active: bool) -> bool {
        CRON_CONTEXT.with(|context| {
            let mut context = context.borrow_mut();
            let previous = context.in_cron_context;
            context.in_cron_context = active;
            previous
        })
    }

    pub fn reset_cron_context(&self, previous: bool) {
        CRON_CONTEXT.with(|context| context.borrow_mut().in_cron_context = previous);
    }
}

impl Tool for CronTool {
    fn name(&self) -> &str {
        "cron"
    }

    fn description(&self) -> &str {
        "Schedule reminders and recurring tasks. Actions: add, list, remove. If tz is omitted, cron expressions and naive ISO times use the tool's default timezone."
    }

    fn parameters(&self) -> Value {
        ToolParameters::new()
            .raw_property(
                "action",
                serde_json::json!({
                    "type": "string",
                    "enum": ["add", "list", "remove"],
                    "description": "Action to perform"
                }),
            )
            .property("name", StringSchema::new("Optional short human-readable label for the job"))
            .property("message", StringSchema::new("Required when action='add'. Instruction for the agent to execute when the job triggers"))
            .property("every_seconds", IntegerSchema::new("Interval in seconds for recurring tasks").minimum(0))
            .property("cron_expr", StringSchema::new("Cron expression like '0 9 * * *'"))
            .property("tz", StringSchema::new("Optional IANA timezone for cron expressions"))
            .property("at", StringSchema::new("ISO datetime for one-time execution"))
            .property("deliver", BooleanSchema::new("Whether to deliver the execution result to the user channel").default(true))
            .property("job_id", StringSchema::new("Required when action='remove'. Job ID to remove"))
            .required(["action"])
            .to_json_schema()
    }

    fn validate_params(&self, params: &JsonMap) -> Vec<ValidationError> {
        let mut errors = crate::tools::base::validate_json_schema_value(
            &Value::Object(params.clone()),
            &self.parameters(),
            "",
        );
        let action = params
            .get("action")
            .and_then(Value::as_str)
            .unwrap_or_default();
        if action == "add"
            && params
                .get("message")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .trim()
                .is_empty()
        {
            errors.push(ValidationError::new(
                "message",
                "is required when action='add'",
            ));
        }
        if action == "remove"
            && params
                .get("job_id")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .trim()
                .is_empty()
        {
            errors.push(ValidationError::new(
                "job_id",
                "is required when action='remove'",
            ));
        }
        errors
    }

    fn execute(&self, params: JsonMap) -> ToolResult {
        let action = params
            .get("action")
            .and_then(Value::as_str)
            .unwrap_or_default();
        match action {
            "add" => self.add_job(params).into(),
            "list" => self.list_jobs().into(),
            "remove" => self
                .remove_job(params.get("job_id").and_then(Value::as_str))
                .into(),
            other => format!("Unknown action: {other}").into(),
        }
    }
}

impl CronTool {
    fn add_job(&self, params: JsonMap) -> String {
        if CRON_CONTEXT.with(|context| context.borrow().in_cron_context) {
            return "Error: cannot schedule new jobs from within a cron job execution".to_owned();
        }
        let message = params
            .get("message")
            .and_then(Value::as_str)
            .unwrap_or_default();
        if message.is_empty() {
            return "Error: cron action='add' requires a non-empty 'message' parameter describing what to do when the job triggers (e.g. the reminder text). Retry including message=\"...\".".to_owned();
        }
        let context = CRON_CONTEXT.with(|context| context.borrow().clone());
        if context.channel.is_empty() || context.chat_id.is_empty() {
            return "Error: no session context (channel/chat_id)".to_owned();
        }
        let cron_expr = params
            .get("cron_expr")
            .and_then(Value::as_str)
            .filter(|value| !value.is_empty());
        let tz = params
            .get("tz")
            .and_then(Value::as_str)
            .filter(|value| !value.is_empty());
        let at = params
            .get("at")
            .and_then(Value::as_str)
            .filter(|value| !value.is_empty());
        if tz.is_some() && cron_expr.is_none() {
            return "Error: tz can only be used with cron_expr".to_owned();
        }

        let (schedule, delete_after_run) = if let Some(every_seconds) = params
            .get("every_seconds")
            .and_then(Value::as_i64)
            .filter(|value| *value > 0)
        {
            (CronSchedule::every(every_seconds * 1000), false)
        } else if let Some(expr) = cron_expr {
            let effective_tz = tz.unwrap_or(&self.default_timezone);
            if let Err(error) = validate_timezone(effective_tz) {
                return error;
            }
            (CronSchedule::cron(expr, effective_tz), false)
        } else if let Some(at) = at {
            match parse_at_ms(at, &self.default_timezone) {
                Ok(at_ms) => (CronSchedule::at(at_ms), true),
                Err(error) => return error,
            }
        } else {
            return "Error: either every_seconds, cron_expr, or at is required".to_owned();
        };

        let name = params
            .get("name")
            .and_then(Value::as_str)
            .filter(|value| !value.is_empty())
            .map(str::to_owned)
            .unwrap_or_else(|| message.chars().take(30).collect());
        let deliver = params
            .get("deliver")
            .and_then(Value::as_bool)
            .unwrap_or(true);
        let job = self.cron.add_job(AddCronJob {
            name,
            schedule,
            message: message.to_owned(),
            deliver,
            channel: Some(context.channel),
            to: Some(context.chat_id),
            delete_after_run,
            channel_meta: context.metadata,
            session_key: if context.session_key.is_empty() {
                None
            } else {
                Some(context.session_key)
            },
        });
        format!("Created job '{}' (id: {})", job.name, job.id)
    }

    fn list_jobs(&self) -> String {
        let jobs = self.cron.list_jobs();
        if jobs.is_empty() {
            return "No scheduled jobs.".to_owned();
        }
        let mut lines = Vec::new();
        for job in jobs {
            let timing = self.format_timing(&job.schedule);
            let mut parts = vec![format!("- {} (id: {}, {timing})", job.name, job.id)];
            if job.payload.kind == CronPayloadKind::SystemEvent {
                parts.push(format!("  Purpose: {}", system_job_purpose(&job)));
                parts
                    .push("  Protected: visible for inspection, but cannot be removed.".to_owned());
            }
            parts.extend(self.format_state(&job.state, &job.schedule));
            lines.push(parts.join("\n"));
        }
        format!("Scheduled jobs:\n{}", lines.join("\n"))
    }

    fn remove_job(&self, job_id: Option<&str>) -> String {
        let Some(job_id) = job_id.filter(|value| !value.is_empty()) else {
            return "Error: job_id is required for remove".to_owned();
        };
        match self.cron.remove_job(job_id) {
            RemoveJobResult::Removed => format!("Removed job {job_id}"),
            RemoveJobResult::Protected => {
                if self
                    .cron
                    .get_job(job_id)
                    .is_some_and(|job| job.name == "dream")
                {
                    "Cannot remove job `dream`.\nThis is a system-managed Dream memory consolidation job for long-term memory.\nIt remains visible so you can inspect it, but it cannot be removed.".to_owned()
                } else {
                    format!("Cannot remove job `{job_id}`.\nThis is a protected system-managed cron job.")
                }
            }
            RemoveJobResult::NotFound => format!("Job {job_id} not found"),
        }
    }

    fn format_timing(&self, schedule: &CronSchedule) -> String {
        match schedule.kind {
            CronScheduleKind::Cron => {
                let tz = schedule
                    .tz
                    .as_deref()
                    .map_or(String::new(), |tz| format!(" ({tz})"));
                format!("cron: {}{tz}", schedule.expr.as_deref().unwrap_or_default())
            }
            CronScheduleKind::Every => format_every(schedule.every_ms.unwrap_or_default()),
            CronScheduleKind::At => format!(
                "at {}",
                format_timestamp(
                    schedule.at_ms.unwrap_or_default(),
                    schedule.tz.as_deref().unwrap_or(&self.default_timezone),
                )
            ),
        }
    }

    fn format_state(&self, state: &CronJobState, schedule: &CronSchedule) -> Vec<String> {
        let display_tz = schedule.tz.as_deref().unwrap_or(&self.default_timezone);
        let mut lines = Vec::new();
        if let Some(last_run_at_ms) = state.last_run_at_ms {
            let status = state
                .last_status
                .as_ref()
                .map(CronRunStatus::as_str)
                .unwrap_or("unknown");
            let mut info = format!(
                "  Last run: {} — {status}",
                format_timestamp(last_run_at_ms, display_tz)
            );
            if let Some(error) = &state.last_error {
                info.push_str(&format!(" ({error})"));
            }
            lines.push(info);
        }
        if let Some(next_run_at_ms) = state.next_run_at_ms {
            lines.push(format!(
                "  Next run: {}",
                format_timestamp(next_run_at_ms, display_tz)
            ));
        }
        lines
    }
}

fn validate_timezone(tz: &str) -> Result<(), String> {
    tz.parse::<Tz>()
        .map(|_| ())
        .map_err(|_| format!("Error: unknown timezone '{tz}'"))
}

fn parse_at_ms(value: &str, default_timezone: &str) -> Result<i64, String> {
    if let Ok(datetime) = DateTime::parse_from_rfc3339(value) {
        return Ok(datetime.timestamp_millis());
    }
    let naive = parse_naive_iso_datetime(value)?;
    let tz = default_timezone
        .parse::<Tz>()
        .map_err(|_| format!("Error: unknown timezone '{default_timezone}'"))?;
    match tz.from_local_datetime(&naive) {
        LocalResult::Single(datetime) => Ok(datetime.timestamp_millis()),
        LocalResult::Ambiguous(earliest, _) => Ok(earliest.timestamp_millis()),
        LocalResult::None => Err(format!(
            "Error: invalid local datetime '{value}' in timezone '{default_timezone}'"
        )),
    }
}

fn format_timestamp(ms: i64, tz_name: &str) -> String {
    let tz = tz_name.parse::<Tz>().unwrap_or(chrono_tz::UTC);
    let formatted = match tz.timestamp_millis_opt(ms) {
        LocalResult::Single(datetime) => datetime.to_rfc3339(),
        LocalResult::Ambiguous(datetime, _) => datetime.to_rfc3339(),
        LocalResult::None => "invalid-timestamp".to_owned(),
    };
    format!("{formatted} ({tz_name})")
}

fn parse_naive_iso_datetime(value: &str) -> Result<NaiveDateTime, String> {
    ["%Y-%m-%dT%H:%M:%S%.f", "%Y-%m-%d %H:%M:%S%.f"]
        .into_iter()
        .find_map(|format| NaiveDateTime::parse_from_str(value, format).ok())
        .ok_or_else(|| {
            format!(
                "Error: invalid ISO datetime format '{value}'. Expected format: YYYY-MM-DDTHH:MM:SS"
            )
        })
}

fn format_every(ms: i64) -> String {
    if ms <= 0 {
        return "every".to_owned();
    }
    if ms % 3_600_000 == 0 {
        format!("every {}h", ms / 3_600_000)
    } else if ms % 60_000 == 0 {
        format!("every {}m", ms / 60_000)
    } else if ms % 1000 == 0 {
        format!("every {}s", ms / 1000)
    } else {
        format!("every {ms}ms")
    }
}

fn system_job_purpose(job: &CronJob) -> &'static str {
    if job.name == "dream" {
        "Dream memory consolidation for long-term memory."
    } else {
        "System-managed internal job."
    }
}
